homework/shirley_homework_2.py

```diff
@@ -0,0 +1,147 @@
+
+employee_dim = build_dimension(full_joined_table, 'employee_id', 'employee_number', ['employee_number', 'last_name', 'first_name', 'reports_to', 'job_title', 'employee_city', 'employee_state', 'employee_country', 'office_code'])
+
+city_dim = build_dimension(full_joined_table, 'city_id', 'order_city', ['order_city', 'order_state', 'order_country'])
+
+
+product_line_dim = full_joined_table[['product_line', 'product_code']].groupby('product_line').count().reset_index()
+product_line_dim.rename(columns={'product_code':'product_line_id'}, inplace=True)
+product_line_dim['product_line_id']= [0,1,2,3,4,5,6]
+
+day_of_week_dim = build_dimension(full_joined_table, 'day_of_week', 'day_id', ['day_id', 'order_date'])
+day_of_week_dim.drop('day_of_week', axis=1, inplace=True)     
+day_of_week_dim['order_date'] = pd.to_datetime(day_of_week_dim['order_date'])
+day_of_week_dim['day'] = day_of_week_dim['order_date'].dt.day_name()
+
+    
+    
+quarter_dim = build_dimension(full_joined_table, 'quarter', 'quarter_id', ['quarter_id', 'order_date'])
+quarter_dim.drop('quarter', axis=1, inplace=True)
+quarter_dim['order_date'] = pd.to_datetime(quarter_dim['order_date'])
+
+#making final star table by merging new columns I created into large merged table,
+#then filtering and only taking the columns I want in the end
+full_joined_table = full_joined_table.merge(day_of_week_dim, on=['order_date','day_id'], how='outer')
+full_joined_table = full_joined_table.merge(quarter_dim, on=['order_date', 'quarter_id'], how='outer')
+full_joined_table = full_joined_table.merge(product_line_dim, on=['product_line'], how='outer')
+full_joined_table = full_joined_table.merge(customer_dim, on=['customer_number'], how='outer')
+full_joined_table = full_joined_table.merge(product_dim, on=['product_code'], how='outer')
+full_joined_table = full_joined_table.merge(city_dim, on='employee_city', how='outer')
+full_joined_table = full_joined_table.merge(employee_dim, on='employee_number', how='outer')
+
+
+cols_to_keep = ['customer_id',\
+                'order_number',\
+                'order_date',\
+                'employee_id',\
+                 'profit',\
+                'total_sale_value',\
+                'day_id',\
+                'quarter_id',\
+                'product_line_id', \
+                'city_id', \
+                'product_id'
+               ]
+
+final_star = full_joined_table[cols_to_keep].reset_index()
+final_star['product_line_id'] = final_star['product_line_id']
+
+
+# Checking for nulls and dropping
+nulls_to_drop = list(final_star[final_star['profit'].isnull()].index)
+nulls_to_drop
+
+final_star.drop(nulls_to_drop, inplace=True)
+
+# Pushing Data to Postgres
+
+
+conn = psycopg2.connect("host=localhost dbname=shirley_homework user=postgres")
+cur = conn.cursor()
+engine = create_engine('postgresql://postgres: @localhost:5432/shirley_homework')
+
+final_star.to_sql(name ='star', con=engine, index=False, if_exists='replace')
```

Comment:
loop?

Reply:
Also, I'm not quite sure what's in this final_star table, but this is a good reason why it can be helpful to CREATE TABLE in SQL first, then add data to it -- that way you can let the database do the work of making sure your data is the way you want it. 

In particular, it seems like your "star" (the fact table) is mixing up "orders" and "order lines", no? As in, `order_number` is not actually unique?